### PR TITLE
DAOS-14442 control: Make NVMe auto-faulty configurable

### PR DESCRIPTION
**Pull request by tanabarr**
SSD auto faulty is enabled by default and criteria set  through
environment variables. Make criteria settable through the server
configuration file instead of environment variables. bdev_auto_faulty
parameter can be used to set enable, max_io_errs and max_csum_errs.

Features: control
Required-githooks: true

### Before requesting gatekeeper:

* [x] Two review approvals and any prior change requests have been resolved.
* [x] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [x] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [x] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by wangshilong — APPROVED** [review]

**Review by wangshilong — DISMISSED** [review]

**Comment by tanabarr** [comment]
I don't think it's worth running through CI again for these indentation check warnings: https://github.com/daos-stack/daos/actions/runs/7454353436/job/20281518852?pr=13548 I will fix in a subsequent PR

**Comment by tanabarr** [comment]
CI run https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-13548/6/tests/ failed with unrelated known issues:
- DFS_Parallel_DTX_cmocka
- EcodOnlineRebuildMdtest
- test_daos_rebuild_ec

**Review by wangshilong — APPROVED** [review]

**Comment by tanabarr** [comment]
CI run DAOS-14845 online_rebuild_mdtest.py Timeout for mdtest after killing one rank

**Review by NiuYawei — APPROVED** [review]
The C changes look good to me.

**Comment by tanabarr** [comment]
CI results for run no. 11 with control and pool features failed for the following known issues:

- https://daosio.atlassian.net/browse/DAOS-14884
- https://daosio.atlassian.net/browse/DAOS-14667

Requesting forced landing

File: utils/config/daos_server.yml [hunk 1]
```diff
@@ -202,25 +202,20 @@
...
-## engines on a host. It needs to represent the total amount of hugepages memory
-## required for all targets across all engines on a host, divided by the system
-## hugepage size. If not set here, it will be automatically calculated based on
-## the number of targets (using the default system hugepage size).
+## Optional parameter that should only be set if overriding the automatically calculated value is #
+## #necessary. Specifies the number (not size) of hugepages to allocate for use by NVMe through
+## #SPDK. For optimum performance each target requires 1 GiB of hugepage space. The provided value
+## should be calculated by dividing the total amount of hugepages memory required for all targets
+## across all engines on a host by the system hugepage size. If not set here, the value will be
+## automatically calculated based on the number of targets (using the default system hugepage size).
```
Comment (knard-intel) [inline utils/config/daos_server.yml:210]:
```suggestion
## Optional parameter that should only be set if overriding the automatically calculated value is 
## necessary. Specifies the number (not size) of hugepages to allocate for use by NVMe through
## SPDK. For optimum performance each target requires 1 GiB of hugepage space. The provided value
## should be calculated by dividing the total amount of hugepages memory required for all targets
## across all engines on a host by the system hugepage size. If not set here, the value will be
## automatically calculated based on the number of targets (using the default system hugepage size).
```

File: src/bio/bio_config.c [hunk 2]
```diff
@@ -936,11 +956,68 @@ bio_read_rpc_srv_settings(const char *nvme_conf, bool *enable, const char **sock
...
+		*max_io_errs = UINT32_MAX;
+	*max_csum_errs = auto_faulty_criteria.max_csum_errs;
+	if (*max_csum_errs == 0)
+		*max_csum_errs = UINT32_MAX;
+
+out:
+	D_INFO("NVMe auto faulty is %s. Criteria: max_io_errs:%u, max_csum_errs:%u\n",
+	       *enable ? "enabled" : "disabled", *max_io_errs, *max_csum_errs);
+
+	if (cfg.method != NULL)
```
Comment (knard-intel) [inline src/bio/bio_config.c:1018]: Nit, not needed to test as D_FREE will do it.

**Review by knard-intel — APPROVED** [review]
Mostly OK for me.
Just minor remarks which could be fix if repush is needed.

File: utils/config/daos_server.yml (hunk 1 above)
Comment (tanabarr) [reply utils/config/daos_server.yml:210]: will fix if repushed

File: src/bio/bio_config.c (hunk 2 above)
Comment (tanabarr) [reply src/bio/bio_config.c:1018]: will fix if repushed

**Comment by mchaarawi** [comment]
since there is a control test failure on the latest run, please merge latest master and add gatekeeper when testing completes.

**Comment by tanabarr** [comment]
clang-format failures to be addressed in a subsequent PR

**Comment by tanabarr** [comment]
> The code changes seem fine to me, but I have some concerns about the design.
> 
> In an ideal world, I think that this configuration would be done as system properties, as I don't see any valid use case for allowing per-engine permutations in behavior. We don't currently have a way to propagate system-level configuration settings out to engines as they join, so I understand that we need to use the server yaml. That having been said, it would probably be better to set these in a top-level entry which is then applied to all engines, similar to how we set the provider and other parameters that should be the same for all engines in a configuration.
> 
> I wouldn't block on making this change, but I recommend that it be made sooner rather than later so that you don't have to support legacy configuration styles.

noted

**Comment by tanabarr** [comment]
@mjmac can you land the PR please?

**Pull request merged**